src/node_quic_crypto.cc

```diff
@@ -0,0 +1,1827 @@
+#include "node_quic_crypto.h"
+#include "env-inl.h"
+#include "node_crypto.h"
+#include "node_quic_session-inl.h"
+#include "node_quic_util.h"
+#include "node_url.h"
+#include "v8.h"
+
+#include <ngtcp2/ngtcp2.h>
+#include <openssl/bio.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/kdf.h>
+#include <openssl/ssl.h>
+#include <openssl/x509v3.h>
+
+#include <iterator>
+#include <numeric>
+#include <unordered_map>
+#include <string>
+#include <sstream>
+
+namespace node {
+
+using v8::Array;
+using v8::Local;
+using v8::Object;
+using v8::String;
+using v8::Value;
+
+namespace quic {
+
+namespace {
+int BIO_Write(BIO* b, const char* buf, int len) {
+  return -1;
+}
+
+int BIO_Read(BIO* b, char* buf, int len) {
+  BIO_clear_retry_flags(b);
+  QuicSession* session = static_cast<QuicSession*>(BIO_get_data(b));
+  len = session->ReadPeerHandshake(reinterpret_cast<uint8_t*>(buf), len);
+  if (len == 0) {
+    BIO_set_retry_read(b);
+    return -1;
+  }
+  return len;
+}
+
+int BIO_Puts(BIO* b, const char* str) {
+  return BIO_Write(b, str, strlen(str));
+}
+
+int BIO_Gets(BIO* b, char* buf, int len) {
+  return -1;
+}
+
+long BIO_Ctrl(  // NOLINT(runtime/int)
+    BIO* b,
+    int cmd,
+    long num,  // NOLINT(runtime/int)
+    void* ptr) {
+  return cmd == BIO_CTRL_FLUSH ? 1 : 0;
+}
+
+int BIO_Create(BIO* b) {
+  BIO_set_init(b, 1);
+  return 1;
+}
+
+int BIO_Destroy(BIO* b) {
+  return b == nullptr ? 0 : 1;
+}
+}  // namespace
+
+BIO_METHOD* CreateBIOMethod() {
+  static BIO_METHOD* method = nullptr;
+
+  if (method == nullptr) {
+    method = BIO_meth_new(BIO_TYPE_FD, "bio");
+    BIO_meth_set_write(method, BIO_Write);
+    BIO_meth_set_read(method, BIO_Read);
+    BIO_meth_set_puts(method, BIO_Puts);
+    BIO_meth_set_gets(method, BIO_Gets);
+    BIO_meth_set_ctrl(method, BIO_Ctrl);
+    BIO_meth_set_create(method, BIO_Create);
+    BIO_meth_set_destroy(method, BIO_Destroy);
+  }
+  return method;
+}
+
+// TODO(@jasnell): Replace with ngtcp2_crypto_aead_keylen once
```

Comment:
The ngtcp2_crypto helper library is part of ngtcp2 and is, for the most part, straightforward. However, there are a couple of functions in ngtcp2_crypto that depend directly on BoringSSL QUIC APIs that are currently being ported to OpenSSL 3. Those are still a WIP and have a copyright issue that still needs to be resolved before we could even begin to rely on them. So, we will either need to (a) wait to adopt ngtcp2_crypto or (b) we float a patch that replaces the BoringSSL APIs with what we have available today or (c) we attempt to port the BoringSSL APIs into the OpenSSL 1.1.1 we have now. I'm currently assuming option (a)